File: BetterTridents/src/main/java/project/kompass/btk/listener/TridentChannelingListener.java
package project.kompass.btk.listener;

import project.kompass.btk.BTK;
import project.kompass.btk.util.TridentUtil;
import org.bukkit.Location;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.event.entity.ProjectileHitEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataType;

public class TridentChannelingListener implements Listener {

    private final BTK plugin;

    public TridentChannelingListener(BTK plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onTridentHit(ProjectileHitEvent event) {
        if (!(event.getEntity() instanceof Trident trident)) return;
        ItemStack item = trident.getItemStack();

        if (item.containsEnchantment(Enchantment.CHANNELING)) {
            if (event.getHitEntity() != null) {
                Entity victim = event.getHitEntity();
                LightningStrike strike = trident.getWorld().spawn(victim.getLocation(), LightningStrike.class);

                strike.getPersistentDataContainer().set(TridentUtil.CHANNELING_LIGHTNING_KEY, PersistentDataType.BYTE, (byte) 1);

                if (trident.getShooter() instanceof Player p) {
                    strike.setCausingPlayer(p);
                }
            } else if (event.getHitBlock() != null) {
                Location loc = event.getHitBlock().getLocation();
                LightningStrike strike = trident.getWorld().spawn(loc, LightningStrike.class);

                strike.getPersistentDataContainer().set(TridentUtil.CHANNELING_LIGHTNING_KEY, PersistentDataType.BYTE, (byte) 1);

                if (trident.getShooter() instanceof Player p) {
                    strike.setCausingPlayer(p);
                }
            }
        }
    }

    @EventHandler
    public void onMeleeHit(EntityDamageByEntityEvent event) {
        if (!(event.getDamager() instanceof Player player)) return;
        ItemStack item = player.getInventory().getItemInMainHand();
        if (item == null || !item.containsEnchantment(Enchantment.CHANNELING)) return;

        if (TridentUtil.isDamageDealingTool(item)) {
            Entity victim = event.getEntity();
            LightningStrike strike = player.getWorld().spawn(victim.getLocation(), LightningStrike.class);

            strike.getPersistentDataContainer().set(TridentUtil.CHANNELING_LIGHTNING_KEY, PersistentDataType.BYTE, (byte) 1);

            strike.setCausingPlayer(player);
        }
    }

    @SuppressWarnings("deprecation")
    @EventHandler
    public void onLightningDamage(EntityDamageByEntityEvent event) {
        if (event.getDamager() instanceof LightningStrike lightning) {
            if (event.getEntity() instanceof Player player) {
                if (player.equals(lightning.getCausingPlayer())) {
                    event.setCancelled(true);
                    return;
                }
            }

            boolean isChanneling = lightning.getPersistentDataContainer().has(
                    TridentUtil.CHANNELING_LIGHTNING_KEY,
                    PersistentDataType.BYTE
            );

            if (isChanneling) {
                event.setDamage(2.0);
            } else {
                event.setDamage(5.0);
            }

            if (event.isApplicable(EntityDamageEvent.DamageModifier.ARMOR)) {
                event.setDamage(EntityDamageEvent.DamageModifier.ARMOR, 0.0);
            }
        }
    }

    @EventHandler
    public void onMobDeath(EntityDeathEvent event) {
        if (!(event.getEntity() instanceof Mob mob)) return;

        boolean killedByChanneling = false;

        if (mob.getLastDamageCause() instanceof EntityDamageByEntityEvent damageEvent) {
            if (damageEvent.getDamager() instanceof Trident trident) {
                if (trident.getItemStack().containsEnchantment(Enchantment.CHANNELING)) {
                    killedByChanneling = true;
                }
            } else if (damageEvent.getDamager() instanceof Player player) {
                ItemStack hand = player.getInventory().getItemInMainHand();
                if (hand != null && hand.containsEnchantment(Enchantment.CHANNELING) && TridentUtil.isDamageDealingTool(hand)) {
                    killedByChanneling = true;
                }
            } else if (damageEvent.getDamager() instanceof LightningStrike strike) {
                if (strike.getCausingPlayer() != null) {
                    killedByChanneling = true;
                }
            }
        }

        if (killedByChanneling) {
            final java.util.List<ItemStack> drops = new java.util.ArrayList<>(event.getDrops());
            final Location deathLoc = mob.getLocation();

            org.bukkit.Bukkit.getScheduler().runTask(plugin, () -> {
                // OPTIMIZATION: Uses Spigot's predicate filter to return ONLY Item drop entities, bypassing nearby mobs/players
                for (Entity entity : deathLoc.getWorld().getNearbyEntities(deathLoc, 1.5, 1.5, 1.5, entity -> entity instanceof Item)) {
                    Item itemEntity = (Item) entity;
                    ItemStack itemStack = itemEntity.getItemStack();
                    for (ItemStack drop : drops) {
                        if (drop.isSimilar(itemStack)) {
                            itemEntity.getPersistentDataContainer().set(
                                    TridentUtil.CHANNELING_PROTECTED_KEY,
                                    PersistentDataType.BYTE,
                                    (byte) 1
                            );
                            break;
                        }
                    }
                }
            });
        }
    }

    @EventHandler
    public void onItemDamage(EntityDamageEvent event) {
        if (event.getEntity() instanceof Item item) {
            if (item.getPersistentDataContainer().has(TridentUtil.CHANNELING_PROTECTED_KEY, PersistentDataType.BYTE)) {
                event.setCancelled(true);
            }
        }
    }
}

File: BetterTridents/src/main/kotlin/project/kompass/btk/listener/TridentChannelingListener.kt
package project.kompass.btk.listener

import project.kompass.btk.BTK
import project.kompass.btk.util.TridentUtil
import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import org.bukkit.Location
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.*
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockIgniteEvent
import org.bukkit.event.entity.*
import org.bukkit.persistence.PersistentDataType
import java.util.concurrent.TimeUnit
import java.util.UUID
import java.util.HashMap

class TridentChannelingListener(private val plugin: BTK) : Listener {

    // Cache tracking ONLY specific channeling mob drops and experience orbs
    private val protectedEntitiesCache: Cache<UUID, Boolean> = CacheBuilder.newBuilder()
        .expireAfterWrite(10, TimeUnit.MINUTES)
        .build()

    private val recentChannelingDeaths = HashMap<Location, Long>()

    // Handles Channeling logic when a thrown trident hits a target
    @EventHandler
    fun onTridentHit(event: ProjectileHitEvent) {
        val trident = event.entity as? Trident ?: return
        val item = trident.itemStack

        if (item.containsEnchantment(Enchantment.CHANNELING)) {
            val hitEntity = event.hitEntity
            val hitBlock = event.hitBlock

            if (hitEntity != null) {
                val strike = trident.world.spawn(hitEntity.location, LightningStrike::class.java)
                strike.persistentDataContainer.set(TridentUtil.CHANNELING_LIGHTNING_KEY, PersistentDataType.BYTE, 1.toByte())
                (trident.shooter as? Player)?.let { strike.causingPlayer = it }
            } else if (hitBlock != null) {
                val strike = trident.world.spawn(hitBlock.location, LightningStrike::class.java)
                strike.persistentDataContainer.set(TridentUtil.CHANNELING_LIGHTNING_KEY, PersistentDataType.BYTE, 1.toByte())
                (trident.shooter as? Player)?.let { strike.causingPlayer = it }
            }
        }
    }

    // Channeling on ANY item during melee attacks
    @EventHandler
    fun onMeleeHit(event: EntityDamageByEntityEvent) {
        if (event.cause == EntityDamageEvent.DamageCause.THORNS) return

        val player = event.damager as? Player ?: return
        val item = player.inventory.itemInMainHand

        if (item.containsEnchantment(Enchantment.CHANNELING)) {
            val victim = event.entity
            val strike = player.world.spawn(victim.location, LightningStrike::class.java)
            strike.persistentDataContainer.set(TridentUtil.CHANNELING_LIGHTNING_KEY, PersistentDataType.BYTE, 1.toByte())
            strike.causingPlayer = player
        }
    }

    // Handles player self-damage cancellation & forces lightning to bypass armor protection
    @Suppress("DEPRECATION")
    @EventHandler
    fun onLightningDamage(event: EntityDamageByEntityEvent) {
        val lightning = event.damager as? LightningStrike ?: return
        val entity = event.entity

        // Prevent player self-damage
        if (entity is Player && entity == lightning.causingPlayer) {
            event.isCancelled = true
            return
        }

        // Note: Catch-all item cancellation removed. Unprotected items (thrown by players) will take damage and burn normally.

        val isChanneling = lightning.persistentDataContainer.has(
            TridentUtil.CHANNELING_LIGHTNING_KEY,
            PersistentDataType.BYTE
        )

        if (isChanneling) {
            event.damage = 2.0
        } else {
            event.damage = 5.0
        }

        if (event.isApplicable(EntityDamageEvent.DamageModifier.ARMOR)) {
            event.setDamage(EntityDamageEvent.DamageModifier.ARMOR, 0.0)
        }
    }

    // Track mob deaths caused by channeling
    @EventHandler
    fun onMobDeath(event: EntityDeathEvent) {
        val mob = event.entity as? Mob ?: return
        var killedByChanneling = false

        val damageEvent = mob.lastDamageCause as? EntityDamageByEntityEvent
        if (damageEvent != null) {
            val damager = damageEvent.damager
            if (damager is Trident) {
                if (damager.itemStack.containsEnchantment(Enchantment.CHANNELING)) {
                    killedByChanneling = true
                }
            } else if (damager is Player) {
                val hand = damager.inventory.itemInMainHand
                if (hand.containsEnchantment(Enchantment.CHANNELING)) {
                    killedByChanneling = true
                }
            } else if (damager is LightningStrike) {
                if (damager.causingPlayer != null) {
                    killedByChanneling = true
                }
            }
        }

        if (killedByChanneling) {
            recentChannelingDeaths[mob.location] = mob.world.fullTime
        }
    }

    // Protects ONLY Item drops and ExperienceOrbs produced by channeling deaths on spawn
    @EventHandler(priority = EventPriority.LOWEST)
    fun onEntitySpawn(event: EntitySpawnEvent) {
        if (recentChannelingDeaths.isEmpty()) return

        val entity = event.entity
        if (entity !is Item && entity !is ExperienceOrb) return

        val loc = entity.location
        val world = entity.world
        val currentTick = world.fullTime

        var isChannelingDrop = false
        val iterator = recentChannelingDeaths.entries.iterator()

        while (iterator.hasNext()) {
            val entry = iterator.next()
            val deathLoc = entry.key
            val deathTick = entry.value

            if (currentTick - deathTick > 10L) {
                iterator.remove()
                continue
            }

            if (deathLoc.world == world && deathLoc.distanceSquared(loc) <= 4.0) {
                isChannelingDrop = true
            }
        }

        if (isChannelingDrop) {
            protectedEntitiesCache.put(entity.uniqueId, true)
            entity.fireTicks = 0
        }
    }

    // Cancel environment damage ONLY for protected mob drops and XP orbs
    @EventHandler
    fun onEntityDamage(event: EntityDamageEvent) {
        val entity = event.entity
        if (entity is Item || entity is ExperienceOrb) {
            if (protectedEntitiesCache.getIfPresent(entity.uniqueId) != null) {
                event.isCancelled = true
                entity.fireTicks = 0
            }
        }
    }

    // Propagate protection to merged item stacks
    @EventHandler
    fun onItemMerge(event: ItemMergeEvent) {
        val target = event.target
        val entity = event.entity

        if (protectedEntitiesCache.getIfPresent(entity.uniqueId) != null ||
            protectedEntitiesCache.getIfPresent(target.uniqueId) != null) {
            protectedEntitiesCache.put(target.uniqueId, true)
        }

        protectedEntitiesCache.invalidate(entity.uniqueId)
    }

    @EventHandler
    fun onItemPickup(event: EntityPickupItemEvent) {
        protectedEntitiesCache.invalidate(event.item.uniqueId)
    }

    @EventHandler
    fun onItemDespawn(event: ItemDespawnEvent) {
        protectedEntitiesCache.invalidate(event.entity.uniqueId)
    }

    // Prevents custom channeling lightning from creating surrounding fire blocks
    @EventHandler
    fun onBlockIgnite(event: BlockIgniteEvent) {
        if (event.cause == BlockIgniteEvent.IgniteCause.LIGHTNING) {
            val igniter = event.ignitingEntity as? LightningStrike ?: return

            if (igniter.persistentDataContainer.has(TridentUtil.CHANNELING_LIGHTNING_KEY, PersistentDataType.BYTE)) {
                event.isCancelled = true
            }
        }
    }
}